// piral-cli/lib/build/run-build-pilet.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("../common");
let handler;
function run(root, piral, sourceMaps, contentHash, minify, externals, importmap, targetDir, outDir, outFile, entryModule, version, logLevel, args) {
    (0, common_1.setStandardEnvs)({
        production: true,
        piral,
        root,
    });
    return handler.create({
        root,
        piral,
        entryModule,
        targetDir,
        outDir,
        outFile,
        externals,
        importmap,
        version,
        develop: false,
        sourceMaps,
        contentHash,
        minify,
        logLevel,
        watch: false,
        args,
    });
}
process.on('message', (msg) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        switch (msg.type) {
            case 'init':
                handler = require(msg.path);
                break;
            case 'start':
                const bundler = yield run(process.cwd(), msg.piral, msg.sourceMaps, msg.contentHash, msg.minify, msg.externals, msg.importmap, msg.targetDir, msg.outDir, msg.outFile, msg.entryModule, msg.version, msg.logLevel, msg);
                const result = yield bundler.bundle();
                if (result) {
                    process.send({
                        type: 'done',
                        outDir: result.outDir,
                        outFile: result.outFile,
                    });
                }
                break;
        }
    }
    catch (error) {
        process.send({
            type: 'fail',
            error: error === null || error === void 0 ? void 0 : error.message,
        });
    }
}));
