// piral-cli/lib/build/run-debug-piral.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const common_1 = require("../common");
let handler;
let bundler;
function run(root, outFile, outDir, piral, hmr, externals, publicUrl, entryFiles, logLevel, args) {
    (0, common_1.setStandardEnvs)({
        root,
        debugPiral: true,
        dependencies: externals,
        publicPath: publicUrl,
        piral,
    });
    return handler.create({
        root,
        entryFiles,
        outFile,
        outDir,
        externals,
        emulator: true,
        sourceMaps: true,
        contentHash: false,
        minify: false,
        publicUrl,
        hmr,
        logLevel,
        watch: true,
        args,
    });
}
process.on('message', (msg) => __awaiter(void 0, void 0, void 0, function* () {
    const root = process.cwd();
    try {
        switch (msg.type) {
            case 'init':
                handler = require(msg.path);
                break;
            case 'bundle':
                if (bundler) {
                    yield bundler.bundle();
                }
                break;
            case 'start':
                const dist = (0, path_1.resolve)(root, 'dist');
                bundler = yield run(root, msg.outFile, dist, msg.piral, msg.hmr, msg.externals, msg.publicUrl, msg.entryFiles, msg.logLevel, msg);
                if (bundler) {
                    bundler.onStart(() => {
                        process.send({
                            type: 'pending',
                        });
                    });
                    bundler.onEnd((result) => {
                        process.send({
                            type: 'update',
                            outHash: result.hash,
                            outName: 'index.html',
                            args: {
                                root,
                            },
                        });
                    });
                    process.send({
                        type: 'done',
                        outDir: dist,
                    });
                }
                break;
        }
    }
    catch (error) {
        process.send({
            type: 'fail',
            error: error === null || error === void 0 ? void 0 : error.message,
        });
    }
}));
