// kras/dist/server/core/recorder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Recorder = void 0;
const uuid_1 = require("uuid");
const events_1 = require("events");
class Recorder extends events_1.EventEmitter {
    constructor(maximum) {
        super();
        this.enabled = true;
        this.requests = [];
        this.errors = [];
        this.messages = [];
        this.maximum = maximum;
    }
    disable() {
        this.enabled = false;
    }
    hit(start, end, request, response) {
        if (this.enabled) {
            const requests = this.requests;
            const id = uuid_1.v4();
            const item = {
                id,
                start,
                end,
                request,
                response,
            };
            if (requests.length === this.maximum) {
                requests.shift();
            }
            this.requests.push(item);
            this.emit('recorded-request', item);
        }
    }
    message(time, data) {
        if (this.enabled) {
            const id = uuid_1.v4();
            const item = Object.assign({ id,
                time }, data);
            this.messages.push(item);
            this.emit('recorded-message', item);
        }
    }
    miss(start, end, request) {
        if (this.enabled) {
            const id = uuid_1.v4();
            const item = {
                id,
                start,
                end,
                request,
            };
            this.errors.push(item);
            this.emit('recorded-miss', item);
        }
    }
}
exports.Recorder = Recorder;
