// kras/dist/server/management/basics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateClient = exports.configOf = void 0;
const path_1 = require("path");
const child_process_1 = require("child_process");
const started = new Date();
function start() {
    const [prog, ...args] = process.argv;
    const cwd = process.cwd();
    child_process_1.spawn(prog, args, {
        cwd,
        detached: true,
        stdio: 'inherit',
    }).unref();
}
function stop() {
    process.exit();
}
function restart() {
    process.on('exit', start);
    stop();
}
function configOf(server, config) {
    const pkgFile = path_1.resolve(__dirname, '..', '..', '..', 'package.json');
    const pkgInfo = require(pkgFile);
    return (_, res) => {
        res.json({
            directory: config.directory,
            sources: config.sources,
            map: config.map,
            name: config.name,
            version: pkgInfo.version,
            started: started.toString(),
            mode: 'running',
        });
    };
}
exports.configOf = configOf;
function updateClient(server, config) {
    return (req, res) => {
        const options = JSON.parse(req.body || '{}');
        switch (options.mode) {
            case 'restart':
                setTimeout(restart, 1000);
                break;
            case 'stop':
                setTimeout(stop, 1000);
                break;
            default:
                break;
        }
        res.sendStatus(200);
    };
}
exports.updateClient = updateClient;
