// piral-cli/lib/common/info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cpuCount = exports.pathSeparator = exports.isWindows = exports.repositoryUrl = exports.compatVersion = exports.cliVersion = exports.cliName = exports.nodeVersion = exports.findCompatVersion = void 0;
const os_1 = require("os");
const info = require('../../package.json');
function findCompatVersion(version) {
    // we only care about major and minor
    const [major, minor] = version.split('.');
    if (major === '0') {
        // we keep the leading zero to avoid confusing, e.g., 0.2 with 2.0
        return `${major}.${minor}`;
    }
    return major;
}
exports.findCompatVersion = findCompatVersion;
exports.nodeVersion = process.version.substring(1);
exports.cliName = info.name;
exports.cliVersion = info.version;
exports.compatVersion = findCompatVersion(exports.cliVersion);
exports.repositoryUrl = info.repository.url;
exports.isWindows = process.platform === 'win32';
exports.pathSeparator = exports.isWindows ? ';' : ':';
exports.cpuCount = (0, os_1.cpus)().length;
