// piral-cli/lib/common/clients/rush.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.listProjects = exports.isProject = exports.initProject = exports.detectClient = exports.installPackage = exports.installDependencies = void 0;
const path_1 = require("path");
const log_1 = require("../log");
const io_1 = require("../io");
const external_1 = require("../../external");
const scripts_1 = require("../scripts");
const MemoryStream_1 = require("../MemoryStream");
// Helpers:
const rushJson = 'rush.json';
function runRushProcess(args, target, output) {
    (0, log_1.log)('generalDebug_0003', 'Starting the Rush process ...');
    const cwd = (0, path_1.resolve)(process.cwd(), target);
    return (0, scripts_1.runCommand)('rush', args, cwd, output);
}
function convert(flags) {
    return flags.map((flag) => {
        switch (flag) {
            case '--save-exact':
                // discard as this may lead to problems
                return '';
            case '--save-dev':
                return '--dev';
            case '--no-save':
                // unfortunately no
                return '';
            default:
                return flag;
        }
    });
}
// Client interface functions:
function installDependencies(target = '.', ...flags) {
    return __awaiter(this, void 0, void 0, function* () {
        const ms = new MemoryStream_1.MemoryStream();
        yield runRushProcess(['update', ...convert(flags)], target, ms);
        (0, log_1.log)('generalDebug_0003', `Rush install dependencies result: ${ms.value}`);
        return ms.value;
    });
}
exports.installDependencies = installDependencies;
function installPackage(packageRef, target = '.', ...flags) {
    return __awaiter(this, void 0, void 0, function* () {
        const ms = new MemoryStream_1.MemoryStream();
        yield runRushProcess(['add', '--package', packageRef, ...convert(flags)], target, ms);
        (0, log_1.log)('generalDebug_0003', `Rush install package result: ${ms.value}`);
        return ms.value;
    });
}
exports.installPackage = installPackage;
function detectClient(root) {
    return __awaiter(this, void 0, void 0, function* () {
        return !!(yield (0, io_1.findFile)(root, rushJson));
    });
}
exports.detectClient = detectClient;
function initProject(packageName, target) {
    return __awaiter(this, void 0, void 0, function* () {
        const rushPath = yield (0, io_1.findFile)(target, rushJson);
        if (!rushPath) {
            throw new Error(`Could not find the "${rushJson}" from "${target}". Sure you want to create the project in the right directory?`);
        }
        const rushDir = (0, path_1.dirname)(rushPath);
        const rushContent = yield (0, io_1.readText)(rushDir, rushJson);
        const rushData = external_1.jju.parse(rushContent);
        const projectFolder = (0, path_1.relative)(rushDir, target);
        if (!Array.isArray(rushData.projects)) {
            rushData.projects = [];
        }
        rushData.projects.push({
            packageName,
            projectFolder,
        });
        yield (0, io_1.writeText)(rushDir, rushJson, external_1.jju.update(rushContent, rushData, {
            mode: 'cjson',
            indent: 2,
        }));
    });
}
exports.initProject = initProject;
function isProject(root, packageRef) {
    var _a, _b;
    return __awaiter(this, void 0, void 0, function* () {
        const details = yield listProjects(root);
        if (typeof details === 'object' && Array.isArray(details === null || details === void 0 ? void 0 : details.projects)) {
            return (_b = (_a = details === null || details === void 0 ? void 0 : details.projects) === null || _a === void 0 ? void 0 : _a.some((p) => p.name === packageRef)) !== null && _b !== void 0 ? _b : false;
        }
        return false;
    });
}
exports.isProject = isProject;
// Functions to exclusively use from rush client:
function listProjects(target) {
    return __awaiter(this, void 0, void 0, function* () {
        const ms = new MemoryStream_1.MemoryStream();
        try {
            yield runRushProcess(['list', '--json'], target, ms);
        }
        catch (e) {
            (0, log_1.log)('generalDebug_0003', `rush list error: ${e}`);
            return {};
        }
        (0, log_1.log)('generalDebug_0003', `rush list project result: ${ms.value}`);
        return JSON.parse(ms.value);
    });
}
exports.listProjects = listProjects;
