// kras/dist/server/info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.injectorMain = exports.injectorConfig = exports.injectorDebug = exports.version = exports.name = exports.author = exports.krasrc = exports.currentDir = exports.rootDir = void 0;
const path_1 = require("path");
const fs_1 = require("fs");
exports.rootDir = path_1.resolve(__dirname, '..', '..');
exports.currentDir = process.cwd();
exports.krasrc = '.krasrc';
function optional(p) {
    try {
        if (fs_1.existsSync(projectFile)) {
            return require(p);
        }
    }
    catch (e) { }
    return undefined;
}
function isInjectorDebug(name) {
    return (/^kras\-[A-Za-z0-9\-]+\-injector$/.test(name) ||
        /^[A-Za-z0-9\-]+\-kras\-injector$/.test(name) ||
        /^[A-Za-z0-9\-]+\-injector$/.test(name));
}
const packageJson = 'package' + '.json';
const packageInfo = require('../../package.json');
const projectFile = path_1.join(exports.currentDir, packageJson);
const projectInfo = optional(projectFile) || {};
exports.author = packageInfo.author;
exports.name = packageInfo.name;
exports.version = packageInfo.version;
exports.injectorDebug = isInjectorDebug(projectInfo.name);
exports.injectorConfig = Object.assign({ active: true }, projectInfo.krasOptions || {});
exports.injectorMain = path_1.resolve(exports.currentDir, projectInfo.main || 'index.js');
