// kras/dist/server/injectors/har-injector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../helpers");
function delay(value, time) {
    if (time) {
        return new Promise((resolve, reject) => {
            setTimeout(() => resolve(value), time);
        });
    }
    return value;
}
function ato(arr) {
    const obj = {};
    for (const item of arr || []) {
        obj[item.name] = item.value;
    }
    return obj;
}
function getUrl(url) {
    if (url.startsWith('http')) {
        return url.substr(url.indexOf('/', 9));
    }
    return url;
}
function findEntries(obj) {
    if (typeof obj === 'object' && obj.log && Array.isArray(obj.log.entries)) {
        return obj.log.entries;
    }
    return [];
}
class HarInjector {
    constructor(options, config) {
        this.files = {};
        const directory = options.directory || config.sources || config.directory;
        this.config = options;
        this.connectors = Object.keys(config.map)
            .filter((target) => config.map[target] !== false)
            .map((target) => ({
            target,
            address: config.map[target],
        }));
        this.watcher = helpers_1.watch(directory, '**/*.har', (ev, fileName) => {
            switch (ev) {
                case 'create':
                case 'update':
                    return this.load(fileName);
                case 'delete':
                    delete this.files[fileName];
                    return;
            }
        });
    }
    getOptions() {
        return {
            delay: {
                description: `If active delays the responses with the time it took according to the HAR.`,
                title: `Delay Responses`,
                type: 'checkbox',
                value: this.config.delay || false,
            },
            directories: helpers_1.editDirectoryOption(this.watcher.directories),
            files: helpers_1.editEntryOption(this.files, ({ request }) => `${request.method} ${request.url}`),
        };
    }
    setOptions(options) {
        this.config.delay = options.delay;
        for (const file of options.files) {
            const entries = this.files[file.name];
            if (entries) {
                for (let i = 0; i < entries.length; i++) {
                    const entry = file.entries[i];
                    if (entry && typeof entry.active === 'boolean') {
                        entries[i].active = entry.active;
                    }
                }
            }
        }
        this.watcher.directories = options.directories;
    }
    get name() {
        return 'har-injector';
    }
    get active() {
        return this.config.active;
    }
    set active(value) {
        this.config.active = value;
    }
    load(fileName) {
        this.files[fileName] = findEntries(helpers_1.asJson(fileName)).map((entry) => this.transformEntry(entry));
    }
    findTarget(url) {
        for (const { target, address } of this.connectors) {
            if (url.indexOf(address) === 0) {
                return target;
            }
        }
        return undefined;
    }
    transformEntry(entry) {
        const original = entry.request;
        const response = entry.response;
        const request = {
            method: original.method,
            url: getUrl(original.url),
            target: original.target || this.findTarget(original.url),
            content: (original.postData || {}).text || '',
            headers: ato(original.headers),
            query: ato(original.queryString),
        };
        delete request.headers._;
        return {
            active: true,
            time: entry.time,
            request,
            response,
        };
    }
    dispose() {
        this.watcher.close();
    }
    handle(req) {
        for (const fileName of Object.keys(this.files)) {
            const entries = this.files[fileName];
            for (let i = 0; i < entries.length; i++) {
                const entry = entries[i];
                if (entry.active) {
                    const item = entry.request;
                    const name = this.name;
                    if (helpers_1.compareRequests(item, req)) {
                        return delay(helpers_1.fromHar(item.url, entry.response, {
                            name,
                            file: {
                                name: fileName,
                                entry: i,
                            },
                        }), this.config.delay && entry.time);
                    }
                }
            }
        }
    }
}
exports.default = HarInjector;
