// kras/dist/server/injectors/json-injector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../helpers");
const faker = require("faker");
const cookie = require("cookie");
const parser = require("accept-language-parser");
const faker_locale_1 = require("../helpers/faker-locale");
const generate_from_template_1 = require("../helpers/generate-from-template");
function find(response, randomize) {
    if (Array.isArray(response)) {
        const index = randomize ? Math.floor(Math.random() * response.length) : 0;
        return response[index];
    }
    return response;
}
class JsonInjector {
    constructor(options, config) {
        this.files = {};
        const directory = options.directory || config.sources || config.directory;
        this.config = options;
        this.watcher = helpers_1.watch(directory, '**/*.json', (ev, fileName) => {
            switch (ev) {
                case 'create':
                case 'update':
                    return this.load(fileName);
                case 'delete':
                    delete this.files[fileName];
                    return;
            }
        });
    }
    getOptions() {
        return {
            randomize: {
                description: `If active randomizes the selected response. Only applicable in case where multiple responses are found for a given request.`,
                title: `Randomize Response`,
                type: 'checkbox',
                value: this.config.randomize || false,
            },
            directories: helpers_1.editDirectoryOption(this.watcher.directories),
            files: helpers_1.editEntryOption(this.files, ({ request }) => `${request.method} ${request.url}`),
        };
    }
    setOptions(options) {
        this.config.randomize = options.randomize;
        for (const file of options.files) {
            const entries = this.files[file.name];
            if (entries) {
                for (let i = 0; i < entries.length; i++) {
                    const entry = file.entries[i];
                    if (entry && typeof entry.active === 'boolean') {
                        entries[i].active = entry.active;
                    }
                }
            }
        }
        this.watcher.directories = options.directories;
    }
    get name() {
        return 'json-injector';
    }
    get active() {
        return this.config.active;
    }
    set active(value) {
        this.config.active = value;
    }
    load(fileName) {
        const content = helpers_1.asJson(fileName);
        const items = Array.isArray(content) ? content : [content];
        for (const item of items) {
            item.active = true;
            if (typeof item.request !== 'object') {
                item.request = {};
            }
            if (typeof item.response !== 'object') {
                item.response = {};
            }
        }
        this.files[fileName] = items;
    }
    dispose() {
        this.watcher.close();
    }
    contentProcess(req, content) {
        if (this.config.generator) {
            const localeName = this.config.generatorLocaleName || 'language';
            const cookies = cookie.parse(req.headers.cookie || '');
            const acceptLanguage = parser.parse(req.headers['accept-language']);
            let locale = 'en';
            if (req.query[localeName]) {
                locale = req.query[localeName];
            }
            else if (cookies[localeName]) {
                locale = cookies[localeName];
            }
            else if (acceptLanguage.length && acceptLanguage[0].code) {
                locale = acceptLanguage[0].code;
            }
            // Convert like: en, en-US to en_US
            faker.setLocale(faker_locale_1.default(locale) || locale);
            // Ignore Buffer content
            if (Buffer.isBuffer(content))
                return content;
            if (typeof content === 'string') {
                content = JSON.parse(content);
            }
            const templateJson = generate_from_template_1.generateFromTemplate(content);
            return JSON.stringify(templateJson);
        }
        return content;
    }
    handle(req) {
        for (const fileName of Object.keys(this.files)) {
            const entries = this.files[fileName];
            for (let i = 0; i < entries.length; i++) {
                const entry = entries[i];
                if (entry.active) {
                    const request = entry.request;
                    if (helpers_1.compareRequests(request, req)) {
                        const rand = this.config.randomize;
                        const response = find(entry.response, rand);
                        const name = this.name;
                        const content = this.contentProcess(req, response.content);
                        return helpers_1.fromJson(request.url, response.status.code, response.status.text, response.headers, content, {
                            name,
                            file: {
                                name: fileName,
                                entry: i,
                            },
                        });
                    }
                }
            }
        }
    }
}
exports.default = JsonInjector;
