// kras/dist/server/injectors/script-injector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tryEvaluate = void 0;
const helpers_1 = require("../helpers");
function errorHandler() {
    return undefined;
}
function tryEvaluate(script) {
    try {
        const handler = helpers_1.asScript(script.file);
        if (typeof handler !== 'function') {
            throw new Error('Does not export a function - it will be ignored.');
        }
        script.error = undefined;
        script.handler = handler;
    }
    catch (e) {
        script.error = e;
        script.handler = errorHandler;
    }
}
exports.tryEvaluate = tryEvaluate;
class ScriptInjector {
    constructor(options, config, core) {
        this.files = {};
        const directory = options.directory || config.sources || config.directory;
        this.config = options;
        this.core = core;
        this.krasConfig = config;
        this.watcher = helpers_1.watch(directory, '**/*.js', (ev, fileName) => {
            switch (ev) {
                case 'create':
                case 'update':
                    return this.load(fileName);
                case 'delete':
                    delete this.files[fileName];
                    return;
            }
        });
    }
    getOptions() {
        return {
            directories: helpers_1.editDirectoryOption(this.watcher.directories),
            files: helpers_1.editFileOption(this.files),
            extended: {
                description: 'The options available to all script files via the context argument.',
                title: 'Extended Configuration',
                type: 'json',
                value: JSON.stringify(this.config.extended || {}, undefined, 2),
            },
        };
    }
    setOptions(options) {
        for (const file of options.files) {
            const script = this.files[file.name];
            const active = file.active;
            if (script && typeof active === 'boolean') {
                script.active = active;
            }
        }
        this.config.extended = JSON.parse(options.extended || '{}');
        this.watcher.directories = options.directories;
    }
    get name() {
        return 'script-injector';
    }
    get active() {
        return this.config.active;
    }
    set active(value) {
        this.config.active = value;
    }
    load(fileName) {
        const script = this.files[fileName] || {
            active: true,
        };
        script.file = fileName;
        tryEvaluate(script);
        if (script.error) {
            this.core.emit('error', script.error);
        }
        this.files[fileName] = script;
    }
    dispose() {
        this.watcher.close();
    }
    handle(req) {
        for (const fileName of Object.keys(this.files)) {
            const script = this.files[fileName];
            const name = this.name;
            if (script.active) {
                const handler = script.handler;
                const builder = ({ statusCode = 200, statusText = '', headers = {}, content = '' }) => helpers_1.fromJson(req.url, statusCode, statusText, headers, content, {
                    name,
                    file: {
                        name: fileName,
                    },
                });
                const extended = this.config.extended || {};
                const ctx = Object.assign({ $server: this.core, $options: this.config, $config: this.krasConfig }, extended);
                const res = handler(ctx, req, builder);
                if (res) {
                    return res;
                }
            }
        }
    }
}
exports.default = ScriptInjector;
