// kras/dist/server/helpers/json-store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.open = exports.JsonStore = void 0;
const io_1 = require("./io");
function all(_item) {
    return true;
}
function none(_item) {
    return false;
}
class JsonStore {
    constructor(file) {
        this.file = file;
    }
    insert(item) {
        const file = this.file;
        const items = io_1.asJson(file) || [];
        items.push(item);
        io_1.toFile(file, items);
    }
    delete(predicate) {
        const file = this.file;
        const items = io_1.asJson(file);
        if (items && items.length) {
            for (let i = items.length; i--;) {
                if ((predicate || none)(items[i])) {
                    items.splice(i, 1);
                }
            }
            io_1.toFile(file, items);
        }
    }
    select(predicate) {
        const file = this.file;
        const items = io_1.asJson(file) || [];
        return items.filter(predicate || all);
    }
    switchTo(file) {
        this.file = file;
    }
}
exports.JsonStore = JsonStore;
function open(file) {
    return new JsonStore(file);
}
exports.open = open;
