// kras/dist/server/helpers/proxy-request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.proxyRequest = exports.defaultProxyHeaders = void 0;
const request = require("request");
const build_response_1 = require("./build-response");
exports.defaultProxyHeaders = [
    'authorization',
    'accept',
    'content-type',
    'cookie',
    'accept-language',
    'user-agent',
    'if-match',
    'if-range',
    'if-unmodified-since',
    'if-none-match',
    'if-modified-since',
    'pragma',
    'range',
];
function proxyRequest(req, callback) {
    var _a;
    return request({
        url: req.url,
        rejectUnauthorized: false,
        method: req.method,
        //tslint:disable-next-line
        encoding: null,
        proxy: req.proxy,
        agentOptions: req.agentOptions,
        headers: req.headers,
        body: req.body,
        followRedirect: (_a = req.redirect) !== null && _a !== void 0 ? _a : true,
    }, (err, ans, body) => {
        if (err) {
            callback(err);
        }
        else {
            callback(undefined, build_response_1.fromNode(ans, body, req.injector));
        }
    });
}
exports.proxyRequest = proxyRequest;
