// piral-cli/lib/commands.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.commands = exports.apps = void 0;
const apps = require("./apps");
exports.apps = apps;
const helpers_1 = require("./helpers");
function specializeCommand(commands, command, suffix) {
    if (command.name.endsWith(suffix)) {
        commands.push(Object.assign(Object.assign({}, command), { name: command.name.replace(suffix, ''), alias: command.alias.filter((n) => n.endsWith(suffix)).map((n) => n.replace(suffix, '')) }));
    }
}
function specializeCommands(suffix) {
    const commands = [];
    for (const command of allCommands) {
        specializeCommand(commands, command, suffix);
    }
    return commands;
}
const allCommands = [
    {
        name: 'debug-piral',
        alias: ['watch-piral', 'debug-portal', 'watch-portal'],
        description: 'Starts the debugging process for a Piral instance.',
        arguments: ['[source]'],
        // "any" due to https://github.com/microsoft/TypeScript/issues/28663 [artifical N = 50]
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source root directory or index.html file for collecting all the information.',
                default: apps.debugPiralDefaults.entry,
            })
                .number('port')
                .describe('port', 'Sets the port of the local development server.')
                .default('port', apps.debugPiralDefaults.port)
                .string('public-url')
                .describe('public-url', 'Sets the public URL (path) of the bundle.')
                .default('public-url', apps.debugPiralDefaults.publicUrl)
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.debugPiralDefaults.logLevel)
                .boolean('open')
                .describe('open', 'Opens the Piral instance directly in the browser.')
                .default('open', apps.debugPiralDefaults.open)
                .boolean('hmr')
                .describe('hmr', 'Activates Hot Module Reloading (HMR).')
                .default('hmr', apps.debugPiralDefaults.hmr)
                .boolean('optimize-modules')
                .describe('optimize-modules', 'Also includes the node modules for target transpilation.')
                .default('optimize-modules', apps.debugPiralDefaults.optimizeModules)
                .choices('bundler', helpers_1.availableBundlers)
                .describe('bundler', 'Sets the bundler to use.')
                .default('bundler', helpers_1.availableBundlers[0])
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.debugPiral(args.base, {
                entry: args.source,
                port: args.port,
                hmr: args.hmr,
                optimizeModules: args['optimize-modules'],
                publicUrl: args['public-url'],
                bundlerName: args.bundler,
                logLevel: args['log-level'],
                open: args.open,
                hooks: args.hooks,
                _: args,
            });
        },
    },
    {
        name: 'build-piral',
        alias: ['bundle-piral', 'build-portal', 'bundle-portal'],
        description: 'Creates a production build for a Piral instance.',
        arguments: ['[source]'],
        // "any" due to https://github.com/microsoft/TypeScript/issues/28663 [artifical N = 50]
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source root directory or index.html file for collecting all the information.',
                default: apps.buildPiralDefaults.entry,
            })
                .string('target')
                .describe('target', 'Sets the target directory or file of bundling.')
                .default('target', apps.buildPiralDefaults.target)
                .string('public-url')
                .describe('public-url', 'Sets the public URL (path) of the bundle. Only for release output.')
                .default('public-url', apps.buildPiralDefaults.publicUrl)
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.buildPiralDefaults.logLevel)
                .boolean('fresh')
                .describe('fresh', 'Performs a fresh build by removing the target directory first.')
                .default('fresh', apps.buildPiralDefaults.fresh)
                .boolean('minify')
                .describe('minify', 'Performs minification or other post-bundle transformations.')
                .default('minify', apps.buildPiralDefaults.minify)
                .boolean('source-maps')
                .describe('source-maps', 'Create associated source maps for the bundles.')
                .default('source-maps', apps.buildPiralDefaults.sourceMaps)
                .boolean('subdir')
                .describe('subdir', `Places the build's output in an appropriate subdirectory (e.g., "emulator"). Ignored for "--all".`)
                .default('subdir', apps.buildPiralDefaults.subdir)
                .boolean('content-hash')
                .describe('content-hash', 'Appends the hash to the side-bundle files.')
                .default('content-hash', apps.buildPiralDefaults.contentHash)
                .boolean('optimize-modules')
                .describe('optimize-modules', 'Also includes the node modules for target transpilation.')
                .default('optimize-modules', apps.buildPiralDefaults.optimizeModules)
                .choices('type', helpers_1.piralBuildTypeKeys)
                .describe('type', 'Selects the target type of the build. "all" builds all target types.')
                .default('type', apps.buildPiralDefaults.type)
                .choices('bundler', helpers_1.availableBundlers)
                .describe('bundler', 'Sets the bundler to use.')
                .default('bundler', helpers_1.availableBundlers[0])
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.buildPiral(args.base, {
                entry: args.source,
                target: args.target,
                publicUrl: args['public-url'],
                bundlerName: args.bundler,
                minify: args.minify,
                fresh: args.fresh,
                subdir: args.subdir,
                contentHash: args['content-hash'],
                sourceMaps: args['source-maps'],
                optimizeModules: args['optimize-modules'],
                logLevel: args['log-level'],
                type: args.type,
                hooks: args.hooks,
                _: args,
            });
        },
    },
    {
        name: 'publish-piral',
        alias: ['release-piral', 'release'],
        description: 'Publishes Piral instance build artifacts.',
        arguments: ['[source]'],
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the previously used output directory to publish.',
                default: apps.publishPiralDefaults.source,
            })
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.publishPiralDefaults.logLevel)
                .choices('type', helpers_1.piralBuildTypeKeys)
                .describe('type', 'Selects the target type to publish. "all" publishes all target types.')
                .default('type', apps.publishPiralDefaults.type)
                .choices('provider', helpers_1.availableReleaseProviders)
                .describe('provider', 'Sets the provider for publishing the release assets.')
                .default('provider', apps.publishPiralDefaults.provider)
                .option('fields', undefined)
                .describe('fields', 'Sets additional fields to be included in the feed service request.')
                .default('fields', apps.publishPiralDefaults.fields)
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.publishPiral(args.base, {
                source: args.source,
                logLevel: args['log-level'],
                type: args.type,
                provider: args.provider,
                fields: args.fields,
            });
        },
    },
    {
        name: 'declaration-piral',
        alias: ['declare-piral', 'declaration-portal', 'declare-portal'],
        description: 'Creates the TypeScript declaration file (index.d.ts) for a Piral instance.',
        arguments: ['[source]'],
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source root directory or index.html file for collecting all the information.',
                default: apps.declarationPiralDefaults.entry,
            })
                .string('target')
                .describe('target', 'Sets the target directory for the generated .d.ts file.')
                .default('target', apps.declarationPiralDefaults.target)
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.declarationPiralDefaults.logLevel)
                .choices('force-overwrite', helpers_1.forceOverwriteKeys)
                .describe('force-overwrite', 'Determines if files should be overwritten by the command.')
                .default('force-overwrite', (0, helpers_1.keyOfForceOverwrite)(apps.declarationPiralDefaults.forceOverwrite))
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.declarationPiral(args.base, {
                entry: args.source,
                target: args.target,
                forceOverwrite: (0, helpers_1.valueOfForceOverwrite)(args['force-overwrite']),
                logLevel: args['log-level'],
            });
        },
    },
    {
        name: 'new-piral',
        alias: ['create-piral', 'scaffold-piral', 'setup-piral'],
        description: 'Creates a new Piral instance by adding all files and changes to the current project.',
        arguments: ['[target]'],
        // "any" due to https://github.com/microsoft/TypeScript/issues/28663 [artifical N = 50]
        flags(argv) {
            return argv
                .positional('target', {
                type: 'string',
                describe: "Sets the project's root directory for making the changes.",
                default: apps.newPiralDefaults.target,
            })
                .string('app')
                .describe('app', "Sets the path to the app's source HTML file.")
                .default('app', apps.newPiralDefaults.app)
                .choices('framework', helpers_1.frameworkKeys)
                .describe('framework', 'Sets the framework/library level to use.')
                .default('framework', apps.newPiralDefaults.framework)
                .boolean('install')
                .describe('install', 'Already performs the installation of its npm dependencies.')
                .default('install', apps.newPiralDefaults.install)
                .string('registry')
                .describe('registry', 'Sets the package registry to use for resolving the dependencies.')
                .default('registry', apps.newPiralDefaults.registry)
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.newPiralDefaults.logLevel)
                .string('tag')
                .describe('tag', 'Sets the tag or version of the package to install. By default, this uses the version of the CLI.')
                .default('tag', apps.newPiralDefaults.version)
                .choices('force-overwrite', helpers_1.forceOverwriteKeys)
                .describe('force-overwrite', 'Determines if files should be overwritten by the installation.')
                .default('force-overwrite', (0, helpers_1.keyOfForceOverwrite)(apps.newPiralDefaults.forceOverwrite))
                .choices('language', helpers_1.piletLanguageKeys)
                .describe('language', 'Determines the programming language for the new Piral instance.')
                .default('language', (0, helpers_1.keyOfPiletLanguage)(apps.newPiralDefaults.language))
                .string('template')
                .describe('template', 'Sets the boilerplate template package to be used when scaffolding.')
                .default('template', apps.newPiralDefaults.template)
                .choices('npm-client', helpers_1.clientTypeKeys)
                .describe('npm-client', 'Sets the npm client to be used when scaffolding.')
                .default('npm-client', apps.newPiralDefaults.npmClient)
                .choices('bundler', helpers_1.bundlerKeys)
                .describe('bundler', 'Sets the default bundler to install.')
                .default('bundler', apps.newPiralDefaults.bundlerName)
                .option('vars', undefined)
                .describe('vars', 'Sets additional variables to be used when scaffolding.')
                .default('vars', apps.newPiralDefaults.variables)
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.newPiral(args.base, {
                app: args.app,
                target: args.target,
                framework: args.framework,
                version: args.tag,
                registry: args.registry,
                forceOverwrite: (0, helpers_1.valueOfForceOverwrite)(args['force-overwrite']),
                language: (0, helpers_1.valueOfPiletLanguage)(args.language),
                install: args.install,
                template: args.template,
                logLevel: args['log-level'],
                npmClient: args['npm-client'],
                bundlerName: args.bundler,
                variables: args.vars,
            });
        },
    },
    {
        name: 'upgrade-piral',
        alias: ['patch'],
        description: 'Upgrades the Piral instance to the latest version of the used Piral packages.',
        arguments: ['[target-version]'],
        flags(argv) {
            return argv
                .positional('target-version', {
                type: 'string',
                describe: 'Sets the tag or version of Piral to upgrade to. By default, it is "latest".',
                default: apps.upgradePiralDefaults.version,
            })
                .string('target')
                .describe('target', 'Sets the target directory to upgrade. By default, the current directory.')
                .default('target', apps.upgradePiralDefaults.target)
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.upgradePiralDefaults.logLevel)
                .boolean('install')
                .describe('install', 'Already performs the update of its npm dependencies.')
                .default('install', apps.upgradePiralDefaults.install)
                .choices('npm-client', helpers_1.clientTypeKeys)
                .describe('npm-client', 'Sets the npm client to be used when upgrading.')
                .default('npm-client', apps.upgradePiralDefaults.npmClient)
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.upgradePiral(args.base, {
                target: args.target,
                version: args['target-version'],
                logLevel: args['log-level'],
                install: args.install,
                npmClient: args['npm-client'],
            });
        },
    },
    {
        name: 'validate-piral',
        alias: ['verify-piral', 'check-piral'],
        description: 'Checks the validity of the current project as a Piral instance.',
        arguments: ['[source]'],
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source root directory or index.html file for collecting all the information.',
                default: apps.validatePiralDefaults.entry,
            })
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.validatePiralDefaults.logLevel)
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.validatePiral(args.base, {
                entry: args.entry,
                logLevel: args['log-level'],
            });
        },
    },
    {
        name: 'debug-pilet',
        alias: ['watch-pilet', 'debug', 'watch'],
        description: 'Starts the debugging process for a pilet using a Piral instance.',
        arguments: ['[source..]'],
        // "any" due to https://github.com/microsoft/TypeScript/issues/28663 [artifical N = 50]
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source file containing the pilet root module.',
                default: apps.debugPiletDefaults.entry,
            })
                .string('target')
                .describe('target', 'Sets the target directory or file of bundling.')
                .default('target', apps.debugPiletDefaults.target)
                .string('public-url')
                .describe('public-url', 'Sets the public URL (path) of the application.')
                .default('public-url', apps.debugPiletDefaults.publicUrl)
                .number('port')
                .describe('port', 'Sets the port of the local development server.')
                .default('port', apps.debugPiletDefaults.port)
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.debugPiletDefaults.logLevel)
                .number('concurrency')
                .describe('concurrency', 'Sets the maximum number of concurrent build jobs.')
                .default('concurrency', apps.debugPiletDefaults.concurrency)
                .boolean('open')
                .describe('open', 'Opens the pilet directly in the browser.')
                .default('open', apps.debugPiletDefaults.open)
                .boolean('hmr')
                .describe('hmr', 'Activates Hot Module Reloading (HMR).')
                .default('hmr', apps.debugPiletDefaults.hmr)
                .boolean('optimize-modules')
                .describe('optimize-modules', 'Also includes the node modules for target transpilation.')
                .default('optimize-modules', apps.debugPiletDefaults.optimizeModules)
                .choices('schema', helpers_1.schemaKeys)
                .describe('schema', 'Sets the schema to be used when bundling the pilets.')
                .default('schema', apps.debugPiletDefaults.schemaVersion)
                .choices('bundler', helpers_1.availableBundlers)
                .describe('bundler', 'Sets the bundler to use.')
                .default('bundler', helpers_1.availableBundlers[0])
                .string('app')
                .describe('app', 'Sets the name of the Piral instance.')
                .string('app-dir')
                .describe('app-dir', 'Sets the path to a custom Piral instance for serving.')
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.')
                .string('feed')
                .describe('feed', 'Sets the URL of a pilet feed for including remote pilets.');
        },
        run(args) {
            return apps.debugPilet(args.base, {
                entry: args.source,
                target: args.target,
                publicUrl: args['public-url'],
                port: args.port,
                hmr: args.hmr,
                bundlerName: args.bundler,
                optimizeModules: args['optimize-modules'],
                appInstanceDir: args['app-dir'],
                app: args.app,
                logLevel: args['log-level'],
                open: args.open,
                schemaVersion: args.schema,
                concurrency: args.concurrency,
                feed: args.feed,
                hooks: args.hooks,
                _: args,
            });
        },
    },
    {
        name: 'build-pilet',
        alias: ['bundle-pilet', 'build', 'bundle'],
        description: 'Creates a production build for a pilet.',
        arguments: ['[source]'],
        // "any" due to https://github.com/microsoft/TypeScript/issues/28663 [artifical N = 50]
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source index.tsx file for collecting all the information.',
                default: apps.buildPiletDefaults.entry,
            })
                .string('target')
                .describe('target', 'Sets the target file of bundling.')
                .default('target', apps.buildPiletDefaults.target)
                .string('public-url')
                .describe('public-url', 'Sets the public URL (path) of the application.')
                .default('public-url', apps.buildPiletDefaults.publicUrl)
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.buildPiletDefaults.logLevel)
                .number('concurrency')
                .describe('concurrency', 'Sets the maximum number of concurrent build jobs.')
                .default('concurrency', apps.buildPiletDefaults.concurrency)
                .boolean('source-maps')
                .describe('source-maps', 'Creates source maps for the bundles.')
                .default('source-maps', apps.buildPiletDefaults.sourceMaps)
                .boolean('fresh')
                .describe('fresh', 'Performs a fresh build by removing the target directory first.')
                .default('fresh', apps.buildPiletDefaults.fresh)
                .boolean('minify')
                .describe('minify', 'Performs minification or other post-bundle transformations.')
                .default('minify', apps.buildPiletDefaults.minify)
                .boolean('declaration')
                .describe('declaration', 'Creates a declaration file for the pilet.')
                .default('declaration', apps.buildPiletDefaults.declaration)
                .boolean('content-hash')
                .describe('content-hash', 'Appends the hash to the side-bundle files.')
                .default('content-hash', apps.buildPiletDefaults.contentHash)
                .boolean('optimize-modules')
                .describe('optimize-modules', 'Also includes the node modules for target transpilation.')
                .default('optimize-modules', apps.buildPiletDefaults.optimizeModules)
                .choices('schema', helpers_1.schemaKeys)
                .describe('schema', 'Sets the schema to be used when bundling the pilets.')
                .default('schema', apps.buildPiletDefaults.schemaVersion)
                .choices('bundler', helpers_1.availableBundlers)
                .describe('bundler', 'Sets the bundler to use.')
                .default('bundler', helpers_1.availableBundlers[0])
                .choices('type', helpers_1.piletBuildTypeKeys)
                .describe('type', 'Selects the target type of the build.')
                .default('type', apps.buildPiletDefaults.type)
                .string('app')
                .describe('app', 'Sets the name of the Piral instance.')
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.buildPilet(args.base, {
                entry: args.source,
                target: args.target,
                publicUrl: args['public-url'],
                minify: args.minify,
                contentHash: args['content-hash'],
                bundlerName: args.bundler,
                declaration: args.declaration,
                sourceMaps: args['source-maps'],
                optimizeModules: args['optimize-modules'],
                type: args.type,
                fresh: args.fresh,
                logLevel: args['log-level'],
                schemaVersion: args.schema,
                concurrency: args.concurrency,
                app: args.app,
                hooks: args.hooks,
                _: args,
            });
        },
    },
    {
        name: 'pack-pilet',
        alias: ['package-pilet', 'pack', 'package'],
        description: 'Creates a pilet package that can be published.',
        arguments: ['[source]'],
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source package.json file for creating the package.',
                default: apps.packPiletDefaults.source,
            })
                .string('target')
                .describe('target', 'Sets the target directory or file of packing.')
                .default('target', apps.packPiletDefaults.target)
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.packPiletDefaults.logLevel)
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.packPilet(args.base, {
                source: args.source,
                target: args.target,
                logLevel: args['log-level'],
            });
        },
    },
    {
        name: 'publish-pilet',
        alias: ['post-pilet', 'publish'],
        description: 'Publishes a pilet package to a pilet feed.',
        arguments: ['[source]'],
        // "any" due to https://github.com/microsoft/TypeScript/issues/28663 [artifical N = 50]
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source of either the previously packed *.tgz bundle or the pilet root module to publish.',
            })
                .string('url')
                .describe('url', 'Sets the explicit URL where to publish the pilet to.')
                .default('url', apps.publishPiletDefaults.url)
                .string('api-key')
                .describe('api-key', 'Sets the potential API key to send to the service.')
                .default('api-key', apps.publishPiletDefaults.apiKey)
                .string('ca-cert')
                .describe('ca-cert', 'Sets a custom certificate authority to use, if any.')
                .default('ca-cert', apps.publishPiletDefaults.cert)
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.publishPiletDefaults.logLevel)
                .boolean('fresh')
                .describe('fresh', 'Performs a fresh build, then packages and finally publishes the pilet.')
                .default('fresh', apps.publishPiletDefaults.fresh)
                .choices('schema', helpers_1.schemaKeys)
                .describe('schema', 'Sets the schema to be used when making a fresh build of the pilet.')
                .default('schema', apps.publishPiletDefaults.schemaVersion)
                .choices('mode', helpers_1.publishModeKeys)
                .describe('mode', 'Sets the authorization mode to use.')
                .default('mode', apps.publishPiletDefaults.mode)
                .choices('bundler', helpers_1.availableBundlers)
                .describe('bundler', 'Sets the bundler to use.')
                .default('bundler', helpers_1.availableBundlers[0])
                .choices('from', helpers_1.fromKeys)
                .describe('from', 'Sets the type of the source to use for publishing.')
                .default('from', apps.publishPiletDefaults.from)
                .option('fields', undefined)
                .describe('fields', 'Sets additional fields to be included in the feed service request.')
                .default('fields', apps.publishPiletDefaults.fields)
                .option('headers', undefined)
                .describe('headers', 'Sets additional headers to be included in the feed service request.')
                .default('headers', apps.publishPiletDefaults.headers)
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.publishPilet(args.base, {
                source: args.source,
                apiKey: args['api-key'],
                url: args.url,
                logLevel: args['log-level'],
                cert: args['ca-cert'],
                bundlerName: args.bundler,
                fresh: args.fresh,
                from: args.from,
                schemaVersion: args.schema,
                fields: args.fields,
                headers: args.headers,
                mode: args.mode,
                _: args,
            });
        },
    },
    {
        name: 'new-pilet',
        alias: ['create-pilet', 'scaffold-pilet', 'scaffold', 'new', 'create'],
        description: 'Scaffolds a new pilet for a specified Piral instance.',
        arguments: ['[source]'],
        // "any" due to https://github.com/microsoft/TypeScript/issues/28663 [artifical N = 50]
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source package (potentially incl. its tag/version) containing a Piral instance for templating the scaffold process.',
                default: apps.newPiletDefaults.source,
            })
                .string('target')
                .describe('target', 'Sets the target directory for scaffolding. By default, the current directory.')
                .default('target', apps.newPiletDefaults.target)
                .string('registry')
                .describe('registry', 'Sets the package registry to use for resolving the specified Piral app.')
                .default('registry', apps.newPiletDefaults.registry)
                .boolean('install')
                .describe('install', 'Already performs the installation of its npm dependencies.')
                .default('install', apps.newPiletDefaults.install)
                .choices('force-overwrite', helpers_1.forceOverwriteKeys)
                .describe('force-overwrite', 'Determines if files should be overwritten by the scaffolding.')
                .default('force-overwrite', (0, helpers_1.keyOfForceOverwrite)(apps.newPiletDefaults.forceOverwrite))
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.newPiletDefaults.logLevel)
                .choices('language', helpers_1.piletLanguageKeys)
                .describe('language', 'Determines the programming language for the new pilet.')
                .default('language', (0, helpers_1.keyOfPiletLanguage)(apps.newPiletDefaults.language))
                .string('template')
                .describe('template', 'Sets the boilerplate template package to be used when scaffolding.')
                .default('template', apps.newPiletDefaults.template)
                .choices('npm-client', helpers_1.clientTypeKeys)
                .describe('npm-client', 'Sets the npm client to be used when scaffolding.')
                .default('npm-client', apps.newPiletDefaults.npmClient)
                .choices('bundler', helpers_1.bundlerKeys)
                .describe('bundler', 'Sets the default bundler to install.')
                .default('bundler', apps.newPiletDefaults.bundlerName)
                .option('vars', undefined)
                .describe('vars', 'Sets additional variables to be used when scaffolding.')
                .default('vars', apps.newPiletDefaults.variables)
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.newPilet(args.base, {
                target: args.target,
                source: args.source,
                registry: args.registry,
                forceOverwrite: (0, helpers_1.valueOfForceOverwrite)(args['force-overwrite']),
                language: (0, helpers_1.valueOfPiletLanguage)(args.language),
                logLevel: args['log-level'],
                install: args.install,
                template: args.template,
                npmClient: args['npm-client'],
                bundlerName: args.bundler,
                variables: args.vars,
            });
        },
    },
    {
        name: 'upgrade-pilet',
        alias: ['upgrade'],
        description: 'Upgrades an existing pilet to the latest version of the used Piral instance.',
        arguments: ['[target-version]'],
        flags(argv) {
            return argv
                .positional('target-version', {
                type: 'string',
                describe: 'Sets the tag or version of the Piral instance to upgrade to. By default, it is "latest".',
                default: apps.upgradePiletDefaults.version,
            })
                .string('target')
                .describe('target', 'Sets the target directory to upgrade. By default, the current directory.')
                .default('target', apps.upgradePiletDefaults.target)
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.upgradePiletDefaults.logLevel)
                .boolean('install')
                .describe('install', 'Already performs the update of its npm dependencies.')
                .default('install', apps.upgradePiletDefaults.install)
                .choices('force-overwrite', helpers_1.forceOverwriteKeys)
                .describe('force-overwrite', 'Determines if files should be overwritten by the upgrading process.')
                .default('force-overwrite', (0, helpers_1.keyOfForceOverwrite)(apps.upgradePiletDefaults.forceOverwrite))
                .choices('npm-client', helpers_1.clientTypeKeys)
                .describe('npm-client', 'Sets the npm client to be used when upgrading.')
                .default('npm-client', apps.upgradePiletDefaults.npmClient)
                .option('vars', undefined)
                .describe('vars', 'Sets additional variables to be used when scaffolding.')
                .default('vars', apps.upgradePiletDefaults.variables)
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.upgradePilet(args.base, {
                target: args.target,
                version: args['target-version'],
                logLevel: args['log-level'],
                forceOverwrite: (0, helpers_1.valueOfForceOverwrite)(args['force-overwrite']),
                install: args.install,
                npmClient: args['npm-client'],
                variables: args.vars,
            });
        },
    },
    {
        name: 'validate-pilet',
        alias: ['verify-pilet', 'check-pilet', 'lint-pilet', 'assert-pilet'],
        description: 'Checks the validity of the current pilet according to the rules defined by the Piral instance.',
        arguments: ['[source]'],
        flags(argv) {
            return argv
                .positional('source', {
                type: 'string',
                describe: 'Sets the source file containing the pilet root module.',
                default: apps.validatePiletDefaults.entry,
            })
                .number('log-level')
                .describe('log-level', 'Sets the log level to use (1-5).')
                .default('log-level', apps.validatePiletDefaults.logLevel)
                .string('app')
                .describe('app', 'Sets the name of the Piral instance.')
                .string('base')
                .default('base', process.cwd())
                .describe('base', 'Sets the base directory. By default the current directory is used.');
        },
        run(args) {
            return apps.validatePilet(args.base, {
                entry: args.entry,
                logLevel: args['log-level'],
                app: args.app,
            });
        },
    },
];
class Commands {
    constructor() {
        this.all = allCommands;
    }
    get pilet() {
        return specializeCommands('-pilet');
    }
    get piral() {
        return specializeCommands('-piral');
    }
}
exports.commands = new Commands();
